aws/asg.go

```diff
@@ -213,67 +206,56 @@ func updateTargetGroupsForAutoScalingGroup(svc autoscalingiface.AutoScalingAPI,
 		}
 
 		if len(detachARNs) > 0 {
-			err = detachTargetGroupsFromAutoScalingGroup(svc, detachARNs, autoScalingGroupName)
+			err := detachTargetGroupsFromAutoScalingGroup(svc, detachARNs, autoScalingGroupName)
 			if err != nil {
 				return err
 			}
 		}
 	}
 
-	// limit target group updates to 10 at a time since this is the limit
-	// in AWS API.
-	chunkSize := 10
-
-	for i := 0; i < len(targetGroupARNs); i += chunkSize {
-		end := i + chunkSize
-
-		if end > len(targetGroupARNs) {
-			end = len(targetGroupARNs)
+	attachARNs := make([]string, 0, len(targetGroupARNs))
+	for _, tgARN := range targetGroupARNs {
+		if _, ok := allTGs[tgARN]; ok {
+			attachARNs = append(attachARNs, tgARN)
+		} else {
+			// TODO: it is better to validate stack's target groups earlier to identify owning stack
+			log.Errorf("target group %q does not exist, will not attach", tgARN)
 		}
+	}
 
-		groups := targetGroupARNs[i:end]
-		if len(groups) > 0 {
-			attachParams := &autoscaling.AttachLoadBalancerTargetGroupsInput{
-				AutoScalingGroupName: aws.String(autoScalingGroupName),
-				TargetGroupARNs:      aws.StringSlice(groups),
-			}
-			_, err = svc.AttachLoadBalancerTargetGroups(attachParams)
-			if err != nil {
-				return err
-			}
+	if len(attachARNs) > 0 {
+		err := attachTargetGroupsToAutoScalingGroup(svc, attachARNs, autoScalingGroupName)
+		if err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
-func describeTagsChunked(svc elbv2iface.ELBV2API, arns []string) ([]*elbv2.TagDescription, error) {
+func describeTags(svc elbv2iface.ELBV2API, arns []string) ([]*elbv2.TagDescription, error) {
 	descs := make([]*elbv2.TagDescription, 0, len(arns))
-	chunkSize := 20
-
-	for i := 0; i < len(arns); i += chunkSize {
-		end := i + chunkSize
-
-		if end > len(arns) {
-			end = len(arns)
-		}
-
-		arnsChunked := arns[i:end]
-		if len(arnsChunked) > 0 {
-			tgParams := &elbv2.DescribeTagsInput{
-				ResourceArns: aws.StringSlice(arnsChunked),
-			}
-
-			tgResp, err := svc.DescribeTags(tgParams)
-			if err != nil {
-				return nil, err
-			}
-
+	err := processChunked(arns, 20, func(chunk []string) error {
+		tgResp, err := svc.DescribeTags(&elbv2.DescribeTagsInput{
+			ResourceArns: aws.StringSlice(chunk),
+		})
+		if err == nil {
 			descs = append(descs, tgResp.TagDescriptions...)
 		}
-	}
+		return err
+	})
+	return descs, err
+}
 
-	return descs, nil
+func describeTargetGroups(elbv2svc elbv2iface.ELBV2API) (map[string]struct{}, error) {
+	targetGroups := make(map[string]struct{})
+	err := elbv2svc.DescribeTargetGroupsPagesWithContext(context.TODO(), &elbv2.DescribeTargetGroupsInput{}, func(resp *elbv2.DescribeTargetGroupsOutput, lastPage bool) bool {
```

Comment:
BTW I am not sure why do we use paged API with context as neither is utilized. There is a simple https://docs.aws.amazon.com/sdk-for-go/api/service/elbv2/#ELBV2.DescribeTargetGroups

Reply:
I might be interpreting it wrong, but IMHO we do use the paging, we are returning always `true` in the paging function, which means we will always get all the content but in multiple chunks. The context is indeed not being used, I believe we planned to use it in the future for easier cancelling (!?). Dunno.

Reply:
I mean we do not use the feature to stop paging early as we always return true so if I do not miss anything the outcome is exactly as using `ELBV2.DescribeTargetGroups`

Reply:
Normally such describe calls in the AWS APIs only list the first `n` (usually around 100) items. From the docs it looks like this one should list all of them, but I would not trust the docs to be right about this as it's usually not the case.

Reply:
yeah, my initial understanding is that `DescribeTargetGroups` would return just the first page.

Reply:
As far as I remember it was exactly the problem we had found: list only the first `n` items, so paging was implemented